namer/namer.cc

```diff
@@ -15,392 +18,730 @@
-        // from enterMethodArgumentSymbol
-        ENFORCE(ctx.owner.data(ctx)->arguments().size() >= pos + 1);
-        unique_ptr<ast::Reference> localExpr = make_unique<ast::Local>(parsedArg.loc, parsedArg.local);
+    u4 idx() const {
+        // Mask top bits.
+        return _id & 0x0FFF;
+    }
 
-        if (parsedArg.default_) {
-            argInfo.flags.isDefault = true;
-            localExpr = ast::MK::OptionalArg(parsedArg.loc, move(localExpr), move(parsedArg.default_));
-        }
+    FoundClassRef &klassRef(FoundNames &foundNames);
+    const FoundClassRef &klassRef(const FoundNames &foundNames) const;
 
-        if (parsedArg.keyword) {
-            argInfo.flags.isKeyword = true;
-        }
-        if (parsedArg.block) {
-            argInfo.flags.isBlock = true;
-        }
-        if (parsedArg.repeated) {
-            argInfo.flags.isRepeated = true;
-        }
+    FoundClass &klass(FoundNames &foundNames);
+    const FoundClass &klass(const FoundNames &foundNames) const;
 
-        return move(localExpr);
+    FoundMethod &method(FoundNames &foundNames);
+    const FoundMethod &method(const FoundNames &foundNames) const;
+
+    FoundFieldOrVariable &fieldOrVariable(FoundNames &foundNames);
+    const FoundFieldOrVariable &fieldOrVariable(const FoundNames &foundNames) const;
+
+    FoundTypeMember &typeMember(FoundNames &foundNames);
+    const FoundTypeMember &typeMember(const FoundNames &foundNames) const;
+
+    core::SymbolRef symbol(const FoundNames &foundNames) const;
+};
+
+struct FoundClassRef final {
+    core::NameRef name;
+    core::LocOffsets loc;
+    // If !owner.exists(), owner is determined by reference site.
+    FoundNameRef owner;
+};
+
+struct FoundClass final {
+    FoundNameRef owner;
+    FoundNameRef klass;
+    core::LocOffsets loc;
+    core::Loc declLoc;
+    ast::ClassDef::Kind classKind;
+};
+
+struct FoundFieldOrVariable final {
+    FoundNameRef owner;
+    FoundNameRef klass;
+    core::NameRef name;
+    core::LocOffsets asgnLoc;
+    core::LocOffsets lhsLoc;
+    bool isTypeAlias = false;
+};
+
+struct FoundTypeMember final {
+    FoundNameRef owner;
+    core::NameRef name;
+    core::LocOffsets asgnLoc;
+    core::LocOffsets nameLoc;
+    core::NameRef varianceName;
+    bool isFixed = false;
+    bool isTypeTemplete = false;
+    bool tooManyArgs = false;
+};
+
+struct FoundMethod final {
+    FoundNameRef owner;
+    core::NameRef name;
+    core::LocOffsets loc;
+    core::Loc declLoc;
+    ast::MethodDef::Flags flags;
+    vector<ast::ParsedArg> parsedArgs;
+};
+
+struct Modifier {
+    NameKind kind;
+    FoundNameRef owner;
+    core::LocOffsets loc;
+    // The name of the modification.
+    core::NameRef name;
+    // For methods: The name of the method being modified.
+    core::NameRef methodName;
+};
+
+class FoundNames final {
+    vector<FoundNameRef> _definitions;
```

Comment:
Could we leave a comment on what these are?

e.g. Aren't _klasses and _methods definitions? ;-)